### PR TITLE
Adds payment API and cart test mule app

### PR DESCRIPTION
**Pull request by jkachel**
### What are the relevant tickets?

https://github.com/mitodl/hq/issues/3166
https://github.com/mitodl/hq/issues/3158

### Description (What does it do?)

Adds the payment API, allowing users to check out via API calls.

Also adds a cart test mule app, which is a very simple frontend that allows users to purchase an item. (Apart from testing, this is a placeholder so that there's a defined space for the actual frontend that this will get; it also holds the interstitial page for redirecting the user into the hosted payment provider.) 

There are also a handful of minor fixes to support the above:
- There is a `docker-compose` change that allows you to specify the port for Postgres by setting `POSTGRES_PORT` in the `.env`. It will default to `5432` - if you have other services, you may want to change that, though.
- The `generate_test_data` command was updated to properly work with revisions for Product objects (it didn't before), and to ensure that the products it creates have prices < $1000 (so you can actually buy them - CyberSource won't authorize a test payment greater than $999). 

### Screenshots (if appropriate):

Screenshot 2024-04-10 at 10 39 59 AM
Screenshot 2024-04-10 at 10 39 50 AM
Screenshot 2024-04-10 at 10 46 15 AM

### How can this be tested?

You will need a working environment set up. This includes Keycloak, and either APISIX or Traefik. See the `README.md` for more details. You should also run `generate_test_data` if you don't already have a system and products set up. 

Automated tests should pass.

Navigating to the homepage - http://ue.odl.local (optionally with a port if you're using APISIX) - should get you a page now, rather than a 404. On that screen, you should be able to choose a product from the dropdown list, and then either add it to cart or add to cart and check out. If you add to cart, you can use the Check Out button to check out and the Clear Cart button should clear your cart successfully. 

Upon checkout, you should see an interstitial page that says "Redirecting to payment processor". The fields for the payload that will be sent to CyberSource are visible here as well as a "Redirect to CyberSource" button at the bottom. You can use this to verify information if you like; the page will automatically redirect in 5 seconds (there's a countdown timer in the button) or you can click the button to skip the timer. Once the redirect happens, you should be sent to CyberSource as per usual and completing the payment workflow there should work as expected. 

Upon successful completion, your cart should be cleared. You should also be able to see that the status of your order has changed to Fulfilled and a Transaction line should be created. There are no notifications on the cart screen at this time (and there likely won't be as this is far from the actual cart interface). No post-sale events will be triggered at this point either - there is future code that depends on this PR that will add that in. 

Cancelling the transaction in CyberSource should work as expected and your cart should remain intact.

Note that unlike other integrations the cart here _does_ support multiple items, so you should be able to check out with an arbitrary number of line items. The items internally link to the system they belong to and (once we have the post-sale events) will trigger actions in those specific systems once payment is complete.

### Additional Context

We'll hide the interstitial form and remove the delay later - it's a bit useful for testing now. (Probably behind a config setting, so it can be toggled back on if necessary.)

The cart app has had no professional styling and does not conform to our standards (for one, it's not a React app). It's solely to have an interface to use for testing. We'll add a real frontend later on but this is not the PR for it. (So, unless there's a dire need to add a feature to it, expanding this specific frontend interface is unlikely to happen.)

Post-sale events don't happen right now - there's a future PR coming up for that.

### CODE REVIEW

File: payments/api.py [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+    Args:
+       order_id (int): ID of the order which is being refunded
+       reference_number (str): Reference number of the order
+       kwargs (dict): Dictionary of the other attributes that are passed e.g.
+       refund amount, refund reason
+       If no refund_amount is provided it will use refund amount from
+       Transaction obj
+
+    Returns:
+        bool : A boolean identifying if an order refund was successful
```
Comment (rhysyngsun) [inline payments/api.py:244]: Return type in docstring doesn't match what's actually returned.

**Review by rhysyngsun — CHANGES_REQUESTED** [review]
Initial pass on review.

**Comment by jkachel** [comment]
This may not be testable if using the APISIX setup locally - I have a backend and a middleware that grab out the data from that (and works) but it's just for DRF endpoints, so getting to the Django Admin or the cart apps won't work. I am working on fixing this. 

Testing locally with Traefik works as that has always just used the normal Django auth backend system.

**Comment by jkachel** [comment]
Last set of commits fixes the auth system so that the frontend parts should work with the APISIX/API7 auth.

File: payments/api.py [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+                cancel_count += 1
+
+                msg = f"Cancelled order {order.reference_number}."
+                log.info(msg)
+            except Exception as e:
+                msg = (
+                    "Couldn't process pending order for cancellation "
+                    f"{payload['req_reference_number']}: {e!s}"
+                )
+                log.exception(msg)
```
Comment (rhysyngsun) [inline payments/api.py:402]: Exception logging should use `%s` + args instead of f-strings.

**Review by rhysyngsun — CHANGES_REQUESTED** [review]
One small code issue

File: payments/api.py (hunk 2 above)
Comment (jkachel) [reply payments/api.py]: fixed in https://github.com/mitodl/unified-ecommerce/pull/99/commits/e5185ab87cc8c8f933cb3b8ddd79d05e497ef77a

**Comment by jkachel** [comment]
Pivoting back into working on this - this is almost ready for review again but I did notice that APISIX mangles the post-CyberSource checkout response route, so need to figure that out. 

This is mostly notes for me: The default setup is a blanket route through APISIX for everything - the checkout result API doesn't/shouldn't require auth though and is POSTed from CyberSource, and the session gets lost in there somewhere, so instead APISIX throws over to Keycloak to restart session, which then drops the payload. So need to add a specific route in APISIX that doesn't do this, then figure out a redirect to display post-sale messaging once it's been processed (and the client has a session again).

**Comment by jkachel** [comment]
Re-requesting review on this since it should be good now.

The routes in APISIX have changed - there's now two, one that's unauthenticated and one that runs requests through Keycloak. If you have the routes set up as they were you need to delete them first. (Or, you can just `docker compose down -v` and start over completely, which honestly might be the easiest route.) 

There's an API endpoint to delete routes (commands here assume you've set your shell up with the API_KEY variable noted in the README): 
1. `curl http://127.0.0.1:9180/apisix/admin/routes/ -H "X-API-KEY: $API_KEY"` to list the routes
2. `curl http://127.0.0.1:9180/apisix/admin/routes/<route name>/ -H "X-API-KEY: $API_KEY" -X DELETE` to delete the routes

You should only have one from the old version of the instructions, so you should just have to run item 2 once. Then you can add the new routes in per the README. The upstream doesn't need to change.

If you opt to start over from scratch, there's a `generate_test_data` management command that will add a bunch of products with random prices in it and their associated records, which puts you pretty close to a usable state right away.

**Review by cp-at-mit — APPROVED** [review]
Looks good!

**Review by rhysyngsun — APPROVED** [review]
LGTM

**Pull request merged**